sparse/core.py

```diff
@@ -603,55 +613,238 @@ def elemwise_binary(self, func, other, *args, **kwargs):
         assert isinstance(other, COO)
         if kwargs.pop('check', True) and func(0, 0, *args, **kwargs) != 0:
             raise ValueError("Performing this operation would produce "
-                    "a dense result: %s" % str(func))
-        if self.shape != other.shape:
-            raise NotImplementedError("Broadcasting is not supported")
+                             "a dense result: %s" % str(func))
+        self_shape, other_shape = self.shape, other.shape
+
+        result_shape = self._get_broadcast_shape(self_shape, other_shape)
+        self_params = self._get_broadcast_parameters(self.shape, result_shape)
+        other_params = self._get_broadcast_parameters(other.shape, result_shape)
+        combined_params = [p1 and p2 for p1, p2 in zip(self_params, other_params)]
+        self_reduce_params = combined_params[-self.ndim:]
+        other_reduce_params = combined_params[-other.ndim:]
```

Comment:
Do any of the extra steps that we do meaningfully affect performance in the common case?

Reply:
I haven't carried out a full benchmark, but asymptotically, no. I'll carry out a full benchmark soon and post the results, hopefully by tomorrow.

Reply:
It seems my version is 2x slower and most of the time is taken by `np.argsort` in the base case. I am considering moving away from `np.recarray`s and onto simple `np.ndarray`s, and from `np.argsort` to `np.lexsort`. This will require me to understand and modify the function `_match_coords` but the rest should follow. Before I do this though, I should probably evaluate the two against each other and actually confirm `np.lexsort` is indeed faster.

Reply:
It seems `np.lexsort` is indeed faster, and by a large margin.
```
import numpy as np
x = np.random.rand(2, 10**7)
x_rec = np.rec.fromarrays(x)
%timeit np.lexsort(x)
4.59 s ± 65.8 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
%timeit np.argsort(x_rec)
10.7 s ± 91.4 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

Reply:
You might want to look at the tricks we do in operations like `linear_loc` to avoid making copies, allocating arrays with dtypes that are as small as possible, and so on.

Reply:
When you get a chance it would be useful to include the examples you're using to benchmark.

Reply:
I've improved the performance quite a bit, now this branch outperforms your `master` branch.

```
import numpy as np
from sparse import COO
from sparse.tests.test_core import random_x
x = COO.from_numpy(random_x((100, 100, 1000)))
y = COO.from_numpy(random_x((100, 100, 1000)))
%timeit x + y

New: 522 ms ± 3.72 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
Old: 1.69 s ± 22.1 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

Reply:
I isolated the performance hit of `np.unique` to only where it's needed.
```
274 ms ± 4.3 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

Reply:
Thanks, it's a pretty useful tool, one that I hadn't seen before. I'm actually a newcomer to Python (only a few months of experience), having fallen in love with its power after leaving MATLAB.

The [PyCharm profiling tool](https://www.jetbrains.com/help/pycharm/optimizing-your-code-using-profilers.html) is pretty useful but it doesn't show stuff in the neat tree format like snakeviz does.

In any case, looks like the function is close to optimal. What should take time intuitively is what is taking time.

Reply:
> Thanks, it's a pretty useful tool, one that I hadn't seen before. I'm actually a newcomer to Python (only a few months of experience), having fallen in love with its power after leaving MATLAB.

Welcome!  I moved over from MatLab myself a long while ago.

Reply:
Latest commit doesn't change code time significantly.